net: renesas: rswitch: add sysfs attribute to enable/disable L3 offload

LKomaryanskiy opened the pull request:
By default, L3 offload is enabled, but in demand to disable it, write 0 to /sys/devices/platform/soc/e68c0000.ethernet/l3_offload file.

LKomaryanskiy committed:
net: renesas: rswitch: add sysfs attribute to enable/disable L3 offload

For some testing scenarios, we need to disable L3 offload. This patch
implements enabling/disabling L3 offload via sysfs attribute for
platform R-Switch device. By default, L3 offload is enabled, but in
case you need to disable it, write 0 to the appropriate sysfs file.
e.g for S4 it placed in
/sys/devices/platform/soc/e68c0000.ethernet/l3_offload

Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

firscity commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -4238,6 +4241,47 @@ static struct notifier_block netevent_notifier = {
+			unregister_fib_notifier(&init_net, &glob_priv->fib_nb);
+		}
+
+		glob_priv->offload_enabled = new_value;
```
I think this should come first and be covered with write_lock, to prevent learning chain from adding new routes after `rswitch_disable_offload`

LKomaryanskiy replied:
It won't happen because after rswitch_disable_offload is executed, the list with routes becomes empty, so learning chain won't add new entries.

firscity commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -4238,6 +4241,47 @@ static struct notifier_block netevent_notifier = {
+		if (new_value) {
+			register_fib_notifier(&init_net, &glob_priv->fib_nb, NULL, NULL);
+		} else {
+			rswitch_disable_offload(glob_priv);
```
Probably these functions should be executed in reverse order.

LKomaryanskiy replied:
Yes, you're right, because it can lead to an issue when we have deleted all routes and in parallel can add new ones because fib_notifier is still registered and works.

firscity commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -1180,7 +1181,7 @@ static bool rswitch_rx_chain(struct net_device *ndev, int *quota, struct rswitch
 		}
 
-		if (learn_chain) {
+		if (learn_chain && rdev->priv->offload_enabled) {
```
Here should be separate if() with read_lock I think

LKomaryanskiy replied:
Thank you for your review.
Actually, this check was added to simplify skipping not needed steps and works slowly but properly without it.
It works fine because rswitch_disable_offload already has all needed locks and after it unlocks, lists with routes for every device become empty. As a result rswitch_add_ipv4_forward doesn't do anything in this case even if offload_enabled flag is true. So, I think, there are no needings to add locks here to protect offload_enabled flag.

firscity reviewed (commented):
Also it will be good to make more detailed commit message: "For some testing cases was needed to..", "Now it is implemented as sysfs flag for platform device..." and remove device address which may be different on other platforms (or at least add "e.g. for S4 SoC it is "/sys/devices..."

firscity reviewed (approved):
`Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>`

LKomaryanskiy commented:
> `Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>`

Thank you, for the close review.

lorc commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -4331,6 +4375,13 @@ static int renesas_eth_sw_probe(struct platform_device *pdev)
+	ret = device_create_file(&pdev->dev, &dev_attr_l3_offload);
+	if (ret) {
+		pr_err("failed to register offload attribute, ret=%d\n", ret);
+		return ret;
```
This error path will leave clocks enabled. Also, it is better to use `dev_err(&pdev->dev, ...)`

LKomaryanskiy replied:
I have reworked error handling in probe and added additional commit for this.

lorc commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -4238,6 +4241,47 @@ static struct notifier_block netevent_notifier = {
+	return count;
+}
+
+static DEVICE_ATTR(l3_offload, 0644, l3_offload_show, l3_offload_store);
```
`s/0644/S_IRUGO | S_IWUSR`

LKomaryanskiy replied:
I changed it according to checkpatch recommendations to DEVICE_ATTR_RW. It's equivalent to S_IRUGO | S_IWUSR or 0644.

lorc commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -4238,6 +4241,47 @@ static struct notifier_block netevent_notifier = {
+	new_value = !!new_value;
+	if (new_value != glob_priv->offload_enabled) {
+		if (new_value) {
+			register_fib_notifier(&init_net, &glob_priv->fib_nb, NULL, NULL);
```
Will this trigger notifiers for existing routing rules?

LKomaryanskiy replied:
Yes, I checked this scenario and after reenabling L3 offload all notifiers triggered properly.

LKomaryanskiy committed:
net: renesas: rswitch: fix freeing resources in case of error probe

Currently, in case of any error in probe, callback just returns
error code and doesn't deinit resources such as disabling clocks,
unregistering notifiers, etc. This patch adds handlers to fix
the issue.

Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: add missing routines in remove and fix deinit ordering

This patch adds missing unregister functions in driver remove callback,
such as unregister_pernet_subsys and unregister_netevent_notifier.
Also, corrected sequence of deinitialization - it should be executed
in reverse order related to initialization.

Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

firscity reviewed (approved):
`Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>`

LKomaryanskiy commented:
@firscity thank you for your close review

lorc commented:
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

LKomaryanskiy commented:
@lorc thank you for your close review. Could I ask you to merge the PR?
